ee/clickhouse/queries/funnels/base.py

```diff
@@ -125,6 +135,42 @@ def _exec_query(self) -> List[Tuple]:
         query = self.get_query()
         return sync_execute(query, self.params)
 
+    def _get_timestamp_outer_select(self) -> str:
+        if self._include_preceding_timestamp:
+            return ", max_timestamp, min_timestamp"
+        elif self._include_timestamp:
+            return ", timestamp"
+        else:
+            return ""
+
+    def _get_timestamp_selects(self) -> Tuple[str, str]:
+        target_step = self._filter.funnel_step
+
+        if not target_step:
+            return "", ""
+
+        if target_step < 0:
+            if target_step == -1:
```

Comment:
Having a hard time parsing what's going on here. A comment to explain please?

Reply:
yeah i'm trying to protect functions more now and hopefully make errors more discoverable but what's happening here is that we use `funnel_step` value to get the proper fields. However, when someone requests a dropoff field `-2` is the first valid field because it refers to people who did the first step and never made it to the second step. `-1` would have no meaning because there's nothing before the first step.

the absolute value calculation below is to account for the indexing we do. We start at 0 in our SQL fields but we start at 1 for funnel_step argument

Reply:
Ah, this is funnels base. Makes sense now, thanks! I was thinking it's in paths 🤦‍♂️